DifferenceGenerator/src/main/kotlin/DifferenceGenerator.kt

```diff
@@ -85,39 +88,112 @@ class DifferenceGenerator(
         encoder.frameRate = 1.0
         encoder.start()
 
-        val video1Images = grabBufferedImages(this.video1Grabber)
-        val video2Images = grabBufferedImages(this.video2Grabber)
+        val video1Hashes = getHashedVideo(video1File)
+        val video2Hashes = getHashedVideo(video2File)
+        val equals = findEquals(video1Hashes, video2Hashes)
+        println("Equals: $equals")
 
-        // execute the alignment algorithm with the images of both videos
-        alignment = algorithm.run(video1Images, video2Images)
+        processAlignment(equals, encoder)
+
+        encoder.stop()
+        encoder.release()
+        this.video1Grabber.stop()
+        this.video2Grabber.stop()
+    }
+
+    private fun processAlignment(
+        equals: ArrayList<Pair<Int, Int>>,
+        encoder: FFmpegFrameRecorder,
+    ) {
+        val video1Length = video1Grabber.lengthInFrames
+        val video2Length = video2Grabber.lengthInFrames
+
+        var nextGrabbedFrame1 = 0
+        var nextGrabbedFrame2 = 0
+
+        // process frames from (0,0) until last match in equals
+        for (i in equals) {
+            val video1Frames = i.first - nextGrabbedFrame1
+            val video2Frames = i.second - nextGrabbedFrame2
+            processFrames(video1Frames, video2Frames, encoder)
+            encoder.record(getColoredFrame(Color.BLACK))
+            alignment += AlignmentElement.MATCH
+            nextGrabbedFrame1 = i.first + 1
+            nextGrabbedFrame2 = i.second + 1
+        }
+
+        // process frames from last match in equals until end of video
+        val video1Frames = video1Length - nextGrabbedFrame1
+        val video2Frames = video2Length - nextGrabbedFrame2
+        processFrames(video1Frames, video2Frames, encoder)
+    }
+
+    private fun getFrames(
+        amount: Int,
+        grabber: FFmpegFrameGrabber,
+    ): ArrayList<BufferedImage> {
+        val images = ArrayList<BufferedImage>()
+        var i = 0
+        while (i < amount) {
+            var frame: Frame? = grabber.grabImage() ?: throw Exception("Video Grabbing calculation is wrong")
```

Comment:
I think that "val" can be used, which makes the non-null assertions unnecessary (? and !! after Frame/frame)